cirq-google/cirq_google/circuit_serializer.py

```diff
@@ -0,0 +1,657 @@
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Support for serializing and deserializing cirq_google.api.v2 protos."""
+
+from itertools import chain
+from typing import (
+    Any,
+    cast,
+    Dict,
+    List,
+    Optional,
+    Union,
+)
+import sympy
+
+import cirq
+from cirq_google import op_deserializer, op_serializer, arg_func_langs
+from cirq_google.api import v2
+from cirq_google.ops import PhysicalZTag
+from cirq_google.ops.calibration_tag import CalibrationTag
+
+
+class CircuitSerializer:
+    """A class for serializing and deserializing programs and operations.
+
+    This class is for serializing cirq_google.api.v2. protos using one
+    message type per gate type.  It serializes qubits by adding a field
+    into the constants table.  Usage is by passing a `cirq.Circuit`
+    to the `serialize()` method of the class, which will produce a
+    `Program` proto.  Likewise, the `deserialize` method will produce
+    a `cirq.Circuit` object from a `Program` proto.
+
+    This class is more performant than the previous `SerializableGateSet`
+    at the cost of some extendability.
+    """
+
+    def __init__(
+        self,
+        gate_set_name: str,
+    ):
+        """Construct the gate set.
+
+        Args:
+            gate_set_name: The name used to identify the gate set.
+            serializers: The OpSerializers to use for serialization.
+                Multiple serializers for a given gate type are allowed and
+                will be checked for a given type in the order specified here.
+                This allows for a given gate type to be serialized into
+                different serialized form depending on the parameters of the
+                gate.
+            deserializers: The OpDeserializers to convert serialized
+                forms of gates or circuits into Operations.
+        """
+        self.gate_set_name = gate_set_name
+
+    def serialize(
+        self,
+        program: cirq.Circuit,
+        msg: Optional[v2.program_pb2.Program] = None,
+        *,
+        arg_function_language: Optional[str] = None,
+        use_constants: bool = True,
+    ) -> v2.program_pb2.Program:
+        """Serialize a Circuit to cirq_google.api.v2.Program proto.
+
+        Args:
+            program: The Circuit to serialize.
+            msg: An optional proto object to populate with the serialization
+                results.
+            arg_function_language: The `arg_function_language` field from
+                `Program.Language`.
+            use_constants: Whether to use constants in serialization. This is
+                required to be True for serializing CircuitOperations.
```

Comment:
What would be the point of allowing this to be false? Could we remove this option and always use the constants table for v2.5?

Reply:
Removed (for now).  My main concern was keeping the function signatures similar to SerializableGateSet, since the next change would be to add an interface that both of them follow, so that I can use both as a "gate_set".  I think I can get around this (or deprecate the use_constants in the other version?  In any case, I will deal with this in the follow-up PR when I unify the interface.